packages/util/src/emulator.ts

```diff
@@ -140,3 +140,114 @@ export function createMockUserToken(
     signature
   ].join('.');
 }
+
+interface EmulatorStatuses {
+  [name: string]: boolean;
+}
+const emulatorStatus: EmulatorStatuses = {};
+
+interface EmulatorSummary {
+  prod: string[];
+  emulator: string[];
+}
+
+// Checks whether any products are running on an emulator
+function getEmulatorSummary(): EmulatorSummary {
+  const summary: EmulatorSummary = {
+    prod: [],
+    emulator: []
+  };
+  for (const key of Object.keys(emulatorStatus)) {
+    if (emulatorStatus[key]) {
+      summary.emulator.push(key);
+    } else {
+      summary.prod.push(key);
+    }
+  }
+  return summary;
+}
+
+function getOrCreateEl(id: string): { created: boolean; element: HTMLElement } {
+  let parentDiv = document.getElementById(id);
+  let created = false;
+  if (!parentDiv) {
+    parentDiv = document.createElement('div');
+    parentDiv.setAttribute('id', id);
+    created = true;
+  }
+  return { created, element: parentDiv };
+}
+
+/**
+ * Updates Emulator Banner. Primarily used for Firebase Studio
+ * @param name
+ * @param isRunningEmulator
+ * @public
+ */
+export function updateEmulatorBanner(
+  name: string,
+  isRunningEmulator: boolean
+): void {
+  if (
+    typeof window === 'undefined' ||
+    typeof document === 'undefined' ||
+    emulatorStatus[name] === isRunningEmulator
+  ) {
+    return;
+  }
+
+  emulatorStatus[name] = isRunningEmulator;
+  const bannerId = '__firebase__banner';
+  const summary = getEmulatorSummary();
+  const showError = summary.prod.length > 0;
+
+  function tearDown(): void {
+    const element = document.getElementById(bannerId);
+    if (element) {
+      element.remove();
+    }
+  }
+
+  function setupDom(): void {
+    const banner = getOrCreateEl(bannerId);
+    const firebaseText: HTMLSpanElement =
+      document.getElementById('__firebase__text') ||
+      document.createElement('span');
+    if (banner.created) {
+      // update styles
+      const bannerEl = banner.element;
+      bannerEl.style.display = 'flex';
+      bannerEl.style.background = '#7faaf0';
+      bannerEl.style.position = 'absolute';
+      bannerEl.style.bottom = '5px';
+      bannerEl.style.left = '5px';
+      bannerEl.style.padding = '.5em';
+      bannerEl.style.borderRadius = '5px';
+      bannerEl.style.alignContent = 'center';
+      const closeBtn = document.createElement('span');
+      closeBtn.style.cursor = 'pointer';
+      closeBtn.style.paddingLeft = '5px';
+      closeBtn.innerHTML = ' &times;';
+      closeBtn.onclick = () => {
+        tearDown();
+      };
+      bannerEl.appendChild(firebaseText);
+      bannerEl.appendChild(closeBtn);
+      document.body.appendChild(bannerEl);
+    }
+    if (showError) {
+      banner.element.style.background = '#cd5c5c';
+      firebaseText.innerText = `Product${
+        summary.prod.length > 0 ? 's' : ''
```

Comment:
this seems pretty hacky, is it too much to add a dependency on something like https://github.com/plurals/pluralize?

Reply:
Not included in the final copy